### ISSUE TITLE
Document whether it is possible to train TorchScript modules

### ISSUE THREAD

**Issue by martinruenz** [pytorch/pytorch#17614]
## 📚 Documentation

I am interested in the following workflow, but my impression is that this use case is not supported (yet?):

1. Prototype a model in **Python**
2. Export model to **C++** via TorchScript (`torch.jit.trace`)
3. Train model in **C++**

Based on the documentation, it is unclear whether this workflow is supported or not. The very first sentence of https://pytorch.org/docs/master/jit.html states

> TorchScript is a way to create serializable and **optimizable models** from PyTorch code

However, it is ambiguous if "**optimizable**" refers to training or the jit compilation process here. 

It seems that `torch::jit::script::Module` is treated as a special case which does not share commonality / a base class with `torch::nn::Module`. Is there also a way to access the parameters of a `torch::jit::script::Module` in order to use it for training? At first I thought "**optimizable models**" was referring to this but I am unsure now.

This issue is related to the post: https://discuss.pytorch.org/t/can-scriptmodule-be-used-for-training/35932

**Comment by zdevito** [comment]
Yeah, thanks for pointing out the ambiguity. The intention of that sentence is that we can optimize the _wall clock performance_ of the model  with the JIT. However, you can also train the script models in C++ or in Python.  In Python training should work normally (ScriptModule is a subclass of nn.Module). In C++, it is not something commonly done yet so the API is not as clean as the python one. E.g. script::Module` has methods for accessing its parameters, but not recursively accumulating all parameter tensors in C++. However, if you did implement that you can pass it to our C++ optimizers and you would be able to train the network in C++. We'll update the doc to make it more clear.

**Comment by martinruenz** [comment]
Thanks for pointing me in the right direction -- accumulating the parameters manually and passing them to an optimizer worked indeed!

**Comment by hhbyyh** [comment]
I'm trying to do something similar, i.e. train the script model in C++. This is probably a stupid question to ask, but is there a way to convert `torch::jit::script::Module` to `torch::nn::Module`? Just to make the training easier.

**Comment by martinruenz** [comment]
I don't think this is possible right now. At least not easily.

**Comment by hhbyyh** [comment]
Thanks @martinruenz . In that case, it seems I can only invoke `get_parameters()` to collect gradients module by module after `loss.backward()`.

Also I want to check if C++ front end of Pytorch can somehow load the Python model, e.g. `torchvision.models.resnet18()`.

**Comment by BynaryCobweb** [comment]
Training torchscript modules with the C++ frontend works fine for me, but it seems like it's impossible to switch between `train()` and `eval()` mode (eg. switch dropout/batch norm on and off) with the current API.

**Comment by pratyush911** [comment]
> Thanks for pointing me in the right direction -- accumulating the parameters manually and passing them to an optimizer worked indeed!

Hey, can you suggest what did you exactly do?

**Comment by hhbyyh** [comment]
Something we did was in https://github.com/hhbyyh/analytics-zoo/blob/pytorchdoc/zoo/src/main/java/com/intel/analytics/zoo/pipeline/api/net/native/com_intel_analytics_zoo_pipeline_api_net_PytorchModel.cpp.

**Comment by iedmrc** [comment]
Hi,
I exactly need the same workflow. What is the current situation of the issue? Is there any tutorial, example, reference or document for that?
Thanks!

**Comment by iedmrc** [comment]
I found this [series](https://discuss.pytorch.org/t/a-series-of-blogs-on-pytorch-c-api-transfer-learning-jupyter-notebook-with-libtorch-xeus-cling-and-more/54628)  composed by @krshrimali

Includes:

- PyTorch C++ API: Installation and MNIST Digit Classification using VGG-16
- PyTorch C++ API: Using Custom Data
- PyTorch C++ API: Using Custom Data to Train a Network
- Classifying Dogs vs Cats using PyTorch C++ API: Part-1
- Classifying Dogs vs Cats using PyTorch C++ API: Part-2
- Applying Transfer Learning on Dogs vs Cats Dataset using PyTorch C++ API
- Setting up Jupyter Notebook (Xeus Cling) for Libtorch and OpenCV Libraries

**Comment by ashkanmy** [comment]
Hi,
I have already trained my model in Python. My model is combination of two other sub-models. I establish my model by calling : 

my_model = ModelAllInOne(num_of_classes,num_of_outputs)

where ModelAllInOne is just a class that puts my two sub-models to gather. Note, I did not mention the sub-models since I believe they have no impact on this question and its probable answers.

However, to export the train model and load in in C++, I say :


num_Of_Classes = 10
num_of_outputs = 200
model_ts = ModelAllInOne(num_Of_Classes,num_of_outputs)
example_ts = torch.rand(1, 3, 48, 160)
traced_script_module = torch.jit.trace(model_ts, example_ts)
traced_script_module.save("traced_model.pt")

Now my question is : If I need to perform the 

                        traced_script_module.save("traced_model.pt")
at any iteration or just do it after the model is trained.

Thanks

**Comment by crissallan** [comment]
> Thanks for pointing me in the right direction -- accumulating the parameters manually and passing them to an optimizer worked indeed!

Hello! May I ask how did you do this step : "accumulating the parameters manually" in C++? Many thanks!

**Comment by BynaryCobweb** [comment]
@crissallan this function gathers all the parameters that require gradients from a torchscript module
```
    void add_parameters(std::shared_ptr<torch::jit::script::Module> module,
                        std::vector<torch::Tensor> &params)
    {
      for (const auto &tensor : module->parameters())
        {
          if (tensor.requires_grad())
            params.push_back(tensor);
        }
      for (auto child : module->children())
        {
          add_parameters(std::make_shared<torch::jit::script::Module>(child),
                         params);
        }
    }
```

**Comment by grinisrit** [comment]
Is it possible to go the other way round in C++ as well: create a `torch::nn::Module` and then convert it to `torch::jit::script::Module` and serialise it? I couldn't find anything equivalent to python's `torch.jit.script()` or `torch.jit.trace()` for C++. That feature would be very helpful to people writing bindings for PyTorch to other languages (usually relying on the C++ frontend).

**Comment by grinisrit** [comment]
What about going the other way round: create and train a model in C++ and export it then to TorchScript. What would be the equivalent of python's `torch.jit.script` or `torch.jit.trace`? That would be helpful for people writing bindings for pytorch to other languages as those typically rely on the C++ frontend.

**Comment by QuanticDisaster** [comment]
Updating the issue as I am interested but I see that :

> Training torchscript modules with the C++ frontend works fine for me, but it seems like it's impossible to switch between `train()` and `eval()` mode (eg. switch dropout/batch norm on and off) with the current API.

So currently it is not possible to train torchscript modules, right ?

**Comment by Bycob** [comment]
> Updating the issue as I am interested but I see that :
> 
> > Training torchscript modules with the C++ frontend works fine for me, but it seems like it's impossible to switch between `train()` and `eval()` mode (eg. switch dropout/batch norm on and off) with the current API.
> 
> So currently it is not possible to train torchscript modules, right ?

`train()` and `eval()` seem to work with `torch.jit.script` now, so you can train a torchscript model obtained using `torch.jit.script`
If you trace the model you can still technically train it, but it's way less flexible.

**Comment by drewm1980** [comment]
@Bycob can you elaborate on "If you trace the model you can still technically train it, but it's way less flexible."?

I'm considering using torch.jit.trace torchscript output for training models, since conversion to that representation has some advantages for what I'm trying to do.

**Comment by Bycob** [comment]
@drewm1980 

> Are you just referring to frozen control flow and device pinning that are inherrent to torch.jit.trace

Exactly. I have already trained traced models and I can say for sure that it works, but you're bound to the fixed flow inherent to tracing. Which means you can't enable / disable dropout, change output size, etc.. and that some more advanced models (e.g. detection models) are not even trainable because they rely on internal control flow.

You can change the device of a traced model as long as it does not contain a hard coded device id, eg. because the model allocates a tensor during its execution.

**Comment by MaratZakirov** [comment]
> Thanks for pointing me in the right direction -- accumulating the parameters manually and passing them to an optimizer worked indeed!

Could you please share code snipet how you transit trainable code from python (torch script) to C++?

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
